File: backend/summarizer.py
from urllib.parse import urlparse
from collections import defaultdict
from datetime import datetime, timedelta, timezone

def safe_host(url: str | None) -> str | None:
    if not url:
        return None
    try:
        return urlparse(url).netloc or None
    except Exception:
        return None

def summarize_events(events: list[dict]) -> list[dict]:
    """
    Returns a list of memory rows to insert:
    [
      { window_start, window_end, url_host, summary_text }
    ]
    Strategy:
      - Group by host
      - Count event types
      - Pull top titles + key UI texts (button text / labels)
    """
    if not events:
        return []

    # Determine time window
    times = [datetime.fromisoformat(e["created_at"].replace("Z", "+00:00")) for e in events if e.get("created_at")]
    window_start = min(times)
    window_end = max(times)

    by_host = defaultdict(list)
    for e in events:
        host = safe_host(e.get("url")) or "unknown"
        by_host[host].append(e)

    memories = []
    for host, host_events in by_host.items():
        type_counts = defaultdict(int)
        titles = []
        ui_texts = []

        for e in host_events:
            type_counts[e.get("type", "unknown")] += 1
            t = e.get("title")
            if t and t not in titles:
                titles.append(t)
            txt = e.get("text_content")
            if txt and txt.strip():
                ui_texts.append(txt.strip())

        top_titles = titles[:3]
        top_texts = []
        for x in ui_texts:
            if x not in top_texts:
                top_texts.append(x)
            if len(top_texts) >= 5:
                break

        summary_lines = []
        summary_lines.append(f"Domain: {host}")
        summary_lines.append("Activity:")
        for k, v in sorted(type_counts.items(), key=lambda kv: -kv[1]):
            summary_lines.append(f"- {k}: {v}")

        if top_titles:
            summary_lines.append(f"Top pages: {', '.join(top_titles)}")

        if top_texts:
            summary_lines.append(f"UI context: {', '.join(top_texts)}")

        summary_text = "\n".join(summary_lines)

        memories.append({
            "window_start": window_start.isoformat(),
            "window_end": window_end.isoformat(),
            "url_host": host,
            "summary_text": summary_text
        })

    return memories

File: supabase/functions/_shared/summarizer.ts
export function safeHost(url: string | null | undefined): string | null {
  if (!url) return null;
  try {
    const u = new URL(url);
    return u.hostname || null;
  } catch {
    return null;
  }
}

interface EventRow {
  created_at?: string;
  url?: string;
  type?: string;
  title?: string;
  text_content?: string;
}

export function summarizeEvents(events: EventRow[]): { window_start: string; window_end: string; url_host: string; summary_text: string }[] {
  if (!events.length) return [];

  const times = events.map((e) => e.created_at).filter(Boolean) as string[];
  const window_start = new Date(Math.min(...times.map((t) => new Date(t).getTime()))).toISOString();
  const window_end = new Date(Math.max(...times.map((t) => new Date(t).getTime()))).toISOString();

  const byHost = new Map<string, EventRow[]>();
  for (const e of events) {
    const host = safeHost(e.url) || "unknown";
    if (!byHost.has(host)) byHost.set(host, []);
    byHost.get(host)!.push(e);
  }

  const memories: { window_start: string; window_end: string; url_host: string; summary_text: string }[] = [];
  for (const [host, hostEvents] of byHost) {
    const typeCounts: Record<string, number> = {};
    const titles: string[] = [];
    const uiTexts: string[] = [];
    for (const e of hostEvents) {
      const t = e.type || "unknown";
      typeCounts[t] = (typeCounts[t] || 0) + 1;
      if (e.title && !titles.includes(e.title)) titles.push(e.title);
      const txt = (e.text_content || "").trim();
      if (txt && !uiTexts.includes(txt)) uiTexts.push(txt);
    }
    const topTitles = titles.slice(0, 3);
    const topTexts = uiTexts.slice(0, 5);
    const sortedTypes = Object.entries(typeCounts).sort((a, b) => b[1] - a[1]);
    const lines = [
      `Domain: ${host}`,
      "Activity:",
      ...sortedTypes.map(([k, v]) => `- ${k}: ${v}`),
      ...(topTitles.length ? [`Top pages: ${topTitles.join(", ")}`] : []),
      ...(topTexts.length ? [`UI context: ${topTexts.join(", ")}`] : []),
    ];
    memories.push({
      window_start,
      window_end,
      url_host: host,
      summary_text: lines.join("\n"),
    });
  }
  return memories;
}
